Rewrite w32-ime-wrap-function-to-control-ime using advice-add

trueroad opened the pull request:
https://github.com/melpa/melpa/pull/7163#issuecomment-703313376
でのご指摘により、 w32-ime-wrap-function-to-control-ime 関数を
`:around` advice による方法で完全に再実装した。

advice-add を使うようにしたため、
Emacs 24.4 以降が必要となる。

引数 _interactive-p, _interactive-arg, _suffix は不要となったが、
wrap-function-to-control-ime 関数との後方互換性確保のため、
ダミーとして受け取る。ダミーなので何を指定しても何の効果もない。

実現方法が変更になったため、
以前の実装に依存した使い方をしていた場合には動作しなくなる可能性がある。
（例：ラップ前後でインタラクティブ性やインタラクティブ引数を変更する、
　ラップ前のオリジナル関数を直接呼ぶ、など。）
しかし、通常はそういった使い方はしていないものと考えられる。

trueroad committed:
Rewrite w32-ime-wrap-function-to-control-ime using advice-add

https://github.com/melpa/melpa/pull/7163#issuecomment-703313376
でのご指摘により、 w32-ime-wrap-function-to-control-ime 関数を
`:around` advice による方法で完全に再実装した。

advice-add を使うようにしたため、
Emacs 24.4 以降が必要となる。

引数 _interactive-p, _interactive-arg, _suffix は不要となったが、
wrap-function-to-control-ime 関数との後方互換性確保のため、
ダミーとして受け取る。ダミーなので何を指定しても何の効果もない。

実現方法が変更になったため、
以前の実装に依存した使い方をしていた場合には動作しなくなる可能性がある。
（例：ラップ前後でインタラクティブ性やインタラクティブ引数を変更する、
　ラップ前のオリジナル関数を直接呼ぶ、など。）
しかし、通常はそういった使い方はしていないものと考えられる。

trueroad commented:
w32-ime.el の wrap-function-to-control-ime ですが、私は

```el
(wrap-function-to-control-ime 'universal-argument t nil)
(wrap-function-to-control-ime 'read-string nil nil)
(wrap-function-to-control-ime 'read-char nil nil)
(wrap-function-to-control-ime 'read-from-minibuffer nil nil)
(wrap-function-to-control-ime 'y-or-n-p nil nil)
(wrap-function-to-control-ime 'yes-or-no-p nil nil)
(wrap-function-to-control-ime 'map-y-or-n-p nil nil)
```

という使い方しかしたことが無く、他の使用例がないかザッとググってみた感じでは、上記のいずれかがほとんどでした。
他の関数に対する使用例も多少ありましたが、上記と同じく、第2引数と第3引数が nil nil になっているか t nil になっているのがほとんどです。

第3引数が nil でなかった使用例で見つかったものは、

```el
(wrap-function-to-control-ime 'YaTeX-typeset-menu t "P")
(wrap-function-to-control-ime 'YaTeX-switch-mode-menu t "P")
```

ぐらいです。第4引数を指定している例は見つかりませんでした。

wrap-function-to-control-ime がやってることは「第1引数で指定した関数を呼び出したとき、自動で IME OFF してから元の関数を実行し、終わったら IME を元に戻すようにラップする」です。
第1引数の関数が interactive だったら第2引数を non-nil にして、第3引数に interactive 引数（interactive 引数なしなら nil)を指定する、
第1引数の関数が非 interactive だったら第2引数、第3引数ともに nil を指定する、
ことで、ラップした関数のインタラクティブ性やインタラクティブ引数が元の関数と同じになるようにしている、
第4引数は元の関数を保存しておくための関数名として使う（nil なら元の関数名に -original を付けた関数として保存する）
と理解しています。

ただこれ、少なくとも Emacs 20 の頃にまで遡れる、かなりレガシーなコードでして、現代的には advice-add を使って第2～4引数不要ででサクッと実現できてます。
advice-add は Emacs 24.4 で導入されたので、現代的には使っても何ら問題ないと思ってます。
後方互換性確保のために少なくともダミーで引数を受け取れるようにしています。
もしかしたらダミー引数をチェックして矛盾していたら（例えば第1引数が interactive 関数なのに第2引数が nil になっているとかその逆とか）警告出力してもいいかもしれないと思いましたが、余計なお世話な気もして、そこまでしていません。

trueroad commented:
本PRをあてたw32-ime.elを[tr-emacs-ime-module](https://github.com/trueroad/tr-emacs-ime-module)と、IMEパッチ環境の両方で軽く動作確認したところ、問題なさそうです。

conao3 commented on w32-ime.el:
```diff
@@ -114,36 +114,26 @@ Even if IME state is not changed, these functiona are maybe called.")
+(defun w32-ime-wrap-advice-around-control-ime (orig-func &rest args)
+  "IME control around calling ORIG-FUNC with ARGS."
+  (when (and (ime-get-mode)
+	     (equal current-input-method "W32-IME"))
```
ここにタブが入っているようです。
.dir-locals.elというファイルを作り、以下のような内容で保存するか、
```
((emacs-lisp-mode
  (indent-tabs-mode . nil)))
```

ファイル末尾に
```
;; Local Variables:
;; indent-tabs-mode: nil
;; End:
```
と記載してスペースでのインデントを強制した方が良いかもしれません。

---

この行の場合は一行で書いても67桁なので一行で書くのも一手だと思います。

trueroad replied:
おっと、インデント強制を入れます。
一応オリジナルはそこを2行で書いているのでとりあえず2行のままにしておきます。

conao3 commented on w32-ime.el:
```diff
@@ -114,36 +114,26 @@ Even if IME state is not changed, these functiona are maybe called.")
+    (ime-force-off))
+  (let ((retval (apply orig-func args)))
+    (when (and (not (ime-get-mode))
+               (equal current-input-method "W32-IME"))
```
一行にするなら、ここも一行にします。

trueroad replied:
上と同じで、ここもとりあえず2行のままにしておきます。

conao3 commented on w32-ime.el:
```diff
@@ -114,36 +114,26 @@ Even if IME state is not changed, these functiona are maybe called.")
+  (when (and (ime-get-mode)
+	     (equal current-input-method "W32-IME"))
+    (ime-force-off))
+  (let ((retval (apply orig-func args)))
```
letでretvalに捕捉していますが、これはprog1で書きかえられると思います。

trueroad replied:
なるほど、そういう方法がありますね。書き換えさせていただきます。

trueroad committed:
Add local variables indent-tabs-mode: nil

スペースのインデントを強制

trueroad committed:
Fix indent

trueroad committed:
Use prog1 instead of let

trueroad commented:
レビューありがとうございます。
9af24e8 の状態でtr-emacs-ime-moduleとIMEパッチと両方で軽く動作確認して問題なさそうです。

conao3 commented:
すいません、prog1で置き換えられるというコメントは一辺では正しいのですが、
動作から考えるとunwind-protectの方が良いのではないかと思いました。
同じように返り値は第1引数の返り値になるので、これもletは必要ないです。

trueroad committed:
Use unwind-protect instead of prog1

trueroad committed:
Fix the case where the original-func changed the IME/IM state on the way

original-func 呼び出しによって IME/IM 状態が変更されていた場合に対応。

IME/IM OFF で呼び出された場合は、original-func 中で状態変更していても、
original-func 後には何もしないようにする。

IME/IM ON で呼び出された場合は、IME/IM OFF してから
original-func を呼び出すが、終了後に状態変更されたか確認して、
必要に応じて IME/IM ON にする。

trueroad commented:
ありがとうございます。
そういえばオリジナルのコードが unwind-protect つかってましたね。
というわけで unwind-protect に書き換えました。
で、さらにオリジナルのコードをよく読んでみると、条件によっては変になりそうだったので、オリジナルの動作にあわせてみました。

conao3 commented:
了解です! よさそうですー!

trueroad commented:
ありがとうございます。マージしました。
